### ISSUE TITLE
Watching for project files external changes

### ISSUE THREAD

**Issue by gorbunkov** [eclipse-che/che#803]
Hi! 
I want to use Che in parallel with another tool that may modify project files source code, say add getters and setter to the class. The problem now is that Che doesn't track that external changes. For example, I have a version 1 of the file opened in Che, then I modify the file with external tool and go back to file editor in Che. Che wil still display version 1 of the file. If I make any change in the file in Che then version 1 + this change will be saved and the changes made by external tool will be lost. 
Maybe you have a solution for this issue? Thanks.

**Comment by TylerJewell** [comment]
@gorbunkov There are a variety of ways to overcome this problem in the future.

Earlier this week, the Che core team made a very large commit on a virtual file system. The essence of the technology is that the Che IDE now has a virtualized layer on top of the physical file system. As a result, you can now do two things as an extension developer:
1. Perform certain project actions over a targeted subset of the file system.  Say analyzing a set of dependencies for a single folder, as opposed to project recursively.
2. Register file system watchers. You can register a watcher for a change to a file, and then trigger an event within the IDE, such as refreshing the editor.  In this example, you can have someone add a new file in the terminal and the project tree widget can get that event and refresh itself.

I believe that the virtual file system has been merged into master, but we have not updated some of the GUI components, such as the project tree automatically refreshing itself.

The file system watchers are available to any extension developer.  @vparfonov  and @gazarenkov  can provide examples of where these can be registered and watched.

Also, separately, @sunix is working on a project using Eclipse Flux that synchronizes editor experiences on different tools. This will give a multi-cursor editing experience such as a developer who is using Eclipse IDE and Eclipse Che on two separate machines, but they can see each other's edits at the same time.

**Comment by gorbunkov** [comment]
Thank you for the answer, @TylerJewell

**Comment by gorbunkov** [comment]
@vparfonov @gazarenkov
Hi!
Can you please provide an example of how I should register a file watcher and trigger the IDE?

**Comment by vparfonov** [comment]
Hi! @gorbunkov 
We are init our default watcher here [ProjectManager](https://github.com/eclipse/che/blob/master/wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/ProjectManager.java#L108-L108)
as you can see it publishes special event type  [VfsWatchEvent](https://github.com/eclipse/che/blob/master/wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/shared/dto/event/VfsWatchEvent.java). Here we have [event broadcaster](https://github.com/eclipse/che/blob/master/wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/notification/VfsWatchBroadcaster.java#L31-L31) it can push event via websocket to the client side, but for now we don't include this class to our standard build.  
So to make it work for you need to: 
1) add to the ProjectApiModule this line: `bind(VfsWatchBroadcaster.class);`
2) subscribe you client on websocket channel "vfs" (see [here](https://github.com/eclipse/che/blob/master/wsagent/che-core-api-project/src/main/java/org/eclipse/che/api/project/shared/dto/event/VfsWatchEvent.java#L24-L24) ) 
To subscribe on client side you need to call 
`messageBus.subscribe(channel, outputHandler)`
We have a lot examples in our code for example [here](https://github.com/eclipse/che/blob/master/plugins/plugin-debugger/che-plugin-debugger-ide/src/main/java/org/eclipse/che/ide/ext/debugger/client/debug/AbstractDebugger.java#L261-L261).
Just pay attention that we have two websocket connections one on _wsmaster_ second on _wsagent_ you need _wsagent_ message bus. To get its instance you need [this](https://github.com/eclipse/che/blob/master/core/platform-api-client-gwt/che-core-client-gwt-machine/src/main/java/org/eclipse/che/api/machine/gwt/client/WsAgentStateController.java#L145-L145).

Hope this helps you. Ask questions if have problems. And thanks for your interest in our product.

**Comment by gorbunkov** [comment]
@vparfonov Thank you very much for a detailed explanation!
